examples/object-oriented-cpp/PineAPPL.hpp

```diff
@@ -0,0 +1,242 @@
+};
+
+/** @brief Coupling powers for each grid. */
+struct Order {
+    /** @brief Exponent of the strong coupling. */
+    uint32_t alphas;
+
+    /** @brief Exponent of the electromagnetic coupling. */
+    uint32_t alpha;
+
+    /** @brief Exponent of the logarithm of the scale factor of the renomalization scale. */
+    uint32_t logxir;
+
+    /** @brief Exponent of the logarithm of the scale factor of the factorization scale. */
+    uint32_t logxif;
+};
+
+struct Grid {
+
+    /** @brief underlying raw object */
+    pineappl_grid *raw;
+
+    /**
+     * @brief constructor
+     * @param lumi luminosity
+     * @param orders orders
+     * @param bin_limits bin limits
+     * @param key_val additional informations
+     */
+    Grid(const Lumi& lumi, const std::vector<Order>& orders, const std::vector<double>& bin_limits, const KeyVal& key_val) {
+        // cast orders
+        const std::size_t n_orders = orders.size();
+        std::vector<std::uint32_t> raw_orders;
+        for (const Order& order : orders) {
+            raw_orders.push_back(order.alphas);
+            raw_orders.push_back(order.alpha);
+            raw_orders.push_back(order.logxir);
+            raw_orders.push_back(order.logxif);
+        }
+        this->raw = pineappl_grid_new(lumi.raw, n_orders, raw_orders.data(), bin_limits.size() - 1, bin_limits.data(), key_val.raw);
+    }
+
+    /** @brief destructor */
+    ~Grid(){ pineappl_grid_delete(this->raw); }
+
+    /**
+     * @brief Number of orders
+     * @return number of orders
+     */
+    std::size_t order_count() const {
+        return pineappl_grid_order_count(this->raw);
+    }
+
+    /**
+     * @brief Number of bins
+     * @return number of bins
+     */
+    std::size_t bin_count() const {
+        return pineappl_grid_bin_count(this->raw);
+    }
+
+    /**
+     * @brief Fill grid for the given parameters
+     * @param x1 first momentum fraction
+     * @param x2 second momentum fraction
+     * @param q2 scale
+     * @param order order index
+     * @param observable observable value
+     * @param lumi luminosity index
+     * @param weight weight
+     */
+    void fill(const double x1, const double x2, const double q2, const std::size_t order, const double observable, const std::size_t lumi, const double weight) const {
+        pineappl_grid_fill(this->raw, x1, x2, q2, order, observable, lumi, weight);
+    }
+
+    /**
+     * @brief perform a convolution of the grid with PDFs
+     * @param pdg_id hadron ID
+     * @param pdf PDF
+     * @param xi_ren renormalization scale variation
+     * @param xi_fac factorization scale variation
+     * @param order_mask order mask
+     * @param lumi_mask luminosity mask
+     * @return prediction for each bin
+     */
+    std::vector<double> convolute_with_one(const std::int32_t pdg_id, LHAPDF::PDF* pdf, const double xi_ren = 1.0, const double xi_fac = 1.0, const std::vector<bool>& order_mask = {}, const std::vector<bool>& lumi_mask = {}) const {
```

Comment:
Pass `pdf` as a reference.

Reply:
I did as requested - but I don't like it much: `pdf` is a pointer in `dyaa.cpp` since lhapdf natively delivers it like this, then I cast it to reference inside `convolute_with_one` just to pass it in the end as pointer to the CAPI; why can't it be pointer all the way (as the beginning and the end request)? Agreed, pointers are potentially dangerous but they are also handy sometimes ... and there is no complicated structure here (it is owned by `dyaa.cpp` from the beginning to the end)

Reply:
From the API point-of-view a reference more clearly expresses that a reference to a valid `PDF` object is required, whereas pointers can be `nullptr`, which could be allowed. This isn't the case here, so a reference should be accepted and everything follows from there.